orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -963,229 +1091,417 @@ common::Status SplitGraph(Graph& graph,
+      // These receives must be kept.
+      continue;
+    }
+    Node* fwd_recv = recv_nodes.at(s);
+    ORT_ENFORCE(fwd_recv);
+    graph.RemoveNode(fwd_recv->Index());
+    // TODO(jufranc): once we enable partition of training graphs, we need to
+    // remove the backward sends too.
   }
 
-  // If the following line is uncommented, middle and last pipeline stages may
-  // have unresolved symbolic shapes. The reason is that some symbolic shapes
-  // are defined for the original inputs, if original inputs are removed, we
-  // lose the hint to resolve symbolic shapes. For example, if an original
-  // input's shape is [batch, sequence, 1024], that input should be provided as
-  // a feed to all pipeline stages. Otherwise, we don't know the actual values
-  // of "batch" and "sequence".
-  //
-  // graph.SetInputs({visited_inputs.begin(), visited_inputs.end()});
-
-  // update the grah with only visited outputs
   graph.SetOutputs({visited_outputs.begin(), visited_outputs.end()});
   graph.SetGraphResolveNeeded();
   graph.SetGraphProtoSyncNeeded();
+  graph.Resolve();
 
-  return graph.Resolve();
+  // TODO(jufranc): once we allow partition of training graphs, we need to add 
+  // some code to make sure that the backward receive starts after the forward
+  // send, or otherwise the computation will get stuck.
+
+  return Status::OK();
 }
 
-Status ApplyPipelinePartitionToMainGraph(
-    Graph& graph,
-    const std::vector<TrainingSession::TrainingConfiguration::CutInfo>& cut_info,
-    size_t pipeline_stage_id,
-    size_t num_pipeline_stage) {
-  size_t split_count = cut_info.size();
-
-  if (num_pipeline_stage != split_count + 1) {
-    return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Wrong pipeline partition cutting info. Total pipeline stage number is ",
-                           num_pipeline_stage,
-                           ", cut info length is: ",
-                           split_count);
+
+Status VerifyAssignment(std::vector<int> stages, int nstages, Graph& graph) {
+
+  // All stages are used.
+  for (int s = 0; s < nstages; ++s) {
+    auto stage_is_used = std::find(std::begin(stages), std::end(stages), s);
+    ORT_RETURN_IF_NOT(stage_is_used != std::end(stages),
+      "Stage " + std::to_string(s) + " was not assigned to any node.");
   }
 
-  std::vector<Node *> send_nodes, recv_nodes;
-  send_nodes.reserve(split_count);
-  recv_nodes.reserve(split_count);
+  // All nodes have been assigned.
+  auto op_assigned = std::find(std::begin(stages), std::end(stages), -1);
+  ORT_RETURN_IF_NOT(op_assigned == std::end(stages), 
+                    "All ops must be assigned to a stage");
+
+  // Edges always go forward.
+  for (size_t i = 0, t = graph.MaxNodeIndex(); i < t; ++i) {
+    Node* node = graph.GetNode(i);
+    int node_stage = stages[i];
+    auto& node_outputs = node->MutableOutputDefs();
+    for (NodeArg* arg : node_outputs) {
+      if (arg == nullptr || !arg->HasTensorOrScalarShape())
```

Comment:
In my experience is always convenient to also check the property NodeArg::Exists before accessing a node arg. See here: https://github.com/microsoft/onnxruntime/pull/5915